sql: EXECUTE stmt doesn't recurse in executor

jordanlewis opened the pull request:
Previously, running a prepared statement via the SQL `EXECUTE` statement
recursed into the executor, resulting in various unintended behavior
such as disabling 1PC transactions.

Now, `EXECUTE` no longer recursively calls the executor. Instead, it
updates the statement being executed to be the prepared statement and
its corresponding substituted placeholder parameters.

In addition, `EXECUTE` is now an explainable_stmt and can be explained
and traced.

This PR contains a refactor to sql.y that introduces a new category of
statement, explainable_stmt, to represent statements that can be
explained, traced, or used as a data source. preparable_stmt, which was
being (ab)used for this purpose, now comprises the 4 standard SQL DML
operators.

This prevents users from preparing DDL statements, which was erroneously
allowed by the SHOW TRACE update, and other statements like SHOW and
HELP.

Fixes #16844.

jordanlewis committed:
sql: EXECUTE stmt doesn't recurse in executor

Previously, running a prepared statement via the SQL `EXECUTE` statement
recursed into the executor, resulting in various unintended behavior
such as disabling 1PC transactions.

Now, `EXECUTE` no longer recursively calls the executor. Instead, it
updates the statement being executed to be the prepared statement and
its corresponding substituted placeholder parameters.

In addition, `EXECUTE` is now an explainable_stmt and can be explained
and traced.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16868)
<!-- Reviewable:end -->

jordanlewis commented:
No new tests here. Is there a way I could check whether 1PC is possible via `EXECUTE`?

andreimatei commented:
For testing, you could use one of the `TestingEvalFilter`s to inspect the kv requests. But what I think you should do is use session tracing and assert something on the trace.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16868#-:-KoJy1sD5wzkk_TgRndK:b-p6cluj)*
<!-- Sent from Reviewable.io -->

jordanlewis committed:
sql/parser: split preparable/explainable stmts

This is a refactor to sql.y that introduces a new category of
statement, explainable_stmt, to represent statements that can be
explained, traced, or used as a data source. preparable_stmt, which was
being (ab)used for this purpose, now comprises the 4 standard SQL DML
operators.

This prevents users from preparing DDL statements, which was erroneously
allowed by the SHOW TRACE update, and other statements like SHOW and
HELP.

jordanlewis commented:
PTAL. I had to do a little refactor to allow `EXECUTE` to be explained and traced. Pulling in @knz as well for a review since it's touching the explain/trace infrastructure a little.

knz commented on pkg/sql/parser/sql.y:
```diff
@@ -1230,17 +1231,21 @@ preparable_stmt:
+  preparable_stmt
+| create_stmt
+| drop_stmt
+| alter_table_stmt
```
Please explain why you are excluding create, drop and alter table from the list of preparable statements?

jordanlewis replied:
Postgres doesn't support DDLs in prepared statements, and I don't see a good reason for us to do so either. DDLs are not parameterizable and in general don't want to get repeated more than once.

knz replied:
Well I heard noises that we either support placeholders for identifiers in unambiguous positions, or that we have a plan to do so. Is there a known reason why this would not-work?

jordanlewis replied:
No known reason why this won't work, but today we don't support placeholders in DDLs.

I'd be happy to see this changed later, but I think that allowing everything to be preparable (in #16524) might have been premature - I probably would have made some noise there if I had noticed what was going on. I think we need more test coverage and thought before we enable preparing for every kind of statement.

knz replied:
:+1: thanks for explaining.

knz reviewed (commented):
The code looks otherwise fine
